crates/bevy_ecs/src/query/state.rs

```diff
@@ -505,10 +520,16 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
             std::any::type_name::<(NewD, NewF)>(), std::any::type_name::<(D, F)>()
         );
 
+        // The dense-ness of the transmuted query depends on the dense-ness of the original filters
+        // and the dense-ness of the new accessed data.
+        let is_dense = NewD::IS_DENSE && NewF::IS_DENSE && self.is_dense;
+
         QueryState {
             world_id: self.world_id,
             archetype_generation: self.archetype_generation,
+            // TODO: This is unsound, since the dense-ness may have changed.
             matched_storage_ids: self.matched_storage_ids.clone(),
+            is_dense,
```

Comment:
Before merging there's this issue to solve. 

This could be done by removing all the `NewD::IS_DENSE && NewF::IS_DENSE &&` checks, thus allowing the static filters to be sparse while iteration still happens in a dense way. This however feels very sketchy and I would prefer to avoid doing this, but it seems to be sound, or at least I can't find an instance that passes the previous checks and then breaks with dense iteration (this does not fully convince me though, the lack of counterexamples doesn't make a theorem true!)

Alternatively we could change the query to iterate over the subsets of archetypes/tables actually matched by `NewD`/`NewF` (fixing #14628 in the process), but this would require an additional `&Archetypes` parameter (which would also suffer from #14629)